### DISCUSSION TITLE
Shapecast only static objects

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#95] [Q&A] [answered]
I have a shapecast that originates from inside of a dynamic body. I would like to ignore all dynamic bodies during the cast and hit only static ones.

Based on the documentation example, I believe I need to create an object layer for the cast, e.g. CAST (2), along with NON_MOVING(0) and MOVING(1). Then map it to the broadphase layer BP_NON_MOVING(0). Then during the query select the NON_MOVING broadphase layer to cast against and would that be CAST object layer for object layer filter?

Would these pseudo steps be correct?
```js
// enable collision for object layer pairs
([
    [ 0, 1 ],
    [ 1, 1 ],
    [ 0, 2 ]
]);

// map object to broadphase layers
({
    0: [ BP_NON_MOVING ],
    1: [ BP_NON_MOVING, BP_MOVING ]
    2: [ BP_NON_MOVING ]
});

// create broadphase filter for cast:
const bpFilter = new Jolt.DefaultBroadPhaseLayerFilter(joltInterface.GetObjectVsBroadPhaseLayerFilter(), BP_NON_MOVING);

// create object filter for cast:
const objFilter = new Jolt.DefaultObjectLayerFilter(joltInterface.GetObjectLayerPairFilter(), 2);

castShape();
```

**Comment by jrouwe** [comment, marked as answer]
> 1: [ BP_NON_MOVING, BP_MOVING ]

I'm not sure I fully understand your pseudo code as this suggests an object layer maps to 2 broadphase layers, but that's not possible, an object layer only ever maps to 1 broadphase layer. This should probably read ```1: [ BP_MOVING ]```.

> const bpFilter = new Jolt.DefaultBroadPhaseLayerFilter(joltInterface.GetObjectVsBroadPhaseLayerFilter(), BP_NON_MOVING);

This function takes an object layer, so BP_NON_MOVING should be replaced by 2.

Other than that I think your code should work. 

Note that instead of creating a layer, you can also use ```BodyFilterJS``` and in the ```ShouldCollideLocked``` function check if ```inBody.IsDynamic()```.

**Comment by LeXXik** [comment]
> I'm not sure I fully understand your pseudo code as this suggests an object layer maps to 2 broadphase layers, but that's not possible, an object layer only ever maps to 1 broadphase layer. This should probably read `1: [ BP_MOVING ]`.

Hmm. I probably got confused by this statement:
```
We also need to implement a ObjectVsBroadPhaseLayerFilter::ShouldCollide that determines which object layer should collide with what broad phase layers, these can be deduced from the two lists above:

NON_MOVING: BP_MOVING, BP_DEBRIS
MOVING: BP_NON_MOVING, BP_MOVING
DEBRIS: BP_NON_MOVING
BULLET: None (these are not simulated so need no collision with other objects)
WEAPON: BP_NON_MOVING, BP_MOVING
```

Where I would then
```js
// for each object layer
{
    // for each bp layer
    {
        bpInterface.MapObjectToBroadPhaseLayer(objectLayer, bpLayer));
    }
}
```

So mapping is actually a 1 to 1 relationship, and I should change it then.

**Reply by jrouwe** [reply]
The ObjectVsBroadPhaseLayerFilter can indeed visit multiple broadphase layers for an object layer. An object layer can colide with multiple other object layers and since each of these layers can map to 1 broadphase layer, this means that a single object layer can collide with multiple broadphase layers.

**Comment by LeXXik** [comment]
Thank you! Your comments helped. Adding the final result for future reference.

```js
// enable collision for object layer pairs
([
    [ 0, 1 ],
    [ 1, 1 ],
    [ 0, 2 ]
]);

// map object to broadphase layers
({
    0: [ BP_NON_MOVING ],
    1: [ BP_MOVING ]
    2: [ BP_NON_MOVING ]
});

// create broadphase filter for cast:
const bpFilter = new Jolt.DefaultBroadPhaseLayerFilter(joltInterface.GetObjectVsBroadPhaseLayerFilter(), 2);

// create object filter for cast:
const objFilter = new Jolt.DefaultObjectLayerFilter(joltInterface.GetObjectLayerPairFilter(), 2);

castShape(); // ignores dynamic objects
```
